File: src/apps/Andriod/app/src/main/java/com/dkellycollins/triggerfinger/data/daos/impl/BitmapDao.java
package com.dkellycollins.triggerfinger.data.daos.impl;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;

import com.dkellycollins.triggerfinger.data.daos.IBitmapDao;
import com.dkellycollins.triggerfinger.util.qa.Assert;

import java.util.HashMap;

public class BitmapDao implements IBitmapDao {

    private final Context _context;
    private final HashMap<Integer, Bitmap> _store;

    public BitmapDao(Context context) {
        Assert.isNotNull(context, "context");

        _context = context;
        _store = new HashMap<>();
    }

    @Override
    public void loadBitmap(int id) {
        Bitmap bitmap = BitmapFactory.decodeResource(_context.getResources(), id);
        _store.put(id, bitmap);
    }

    @Override
    public Bitmap retrieve(int id) {
        return _store.get(id);
    }

    @Override
    public void delete(int id) {
        Bitmap bitmap = _store.get(id);

        bitmap.recycle();

        _store.remove(id);
    }
}

File: src/apps/Android-Kotlin/app/src/main/kotlin/com/dkellycollins/triggerfinger/data/daos/impl/BitmapDao.kt
package com.dkellycollins.triggerfinger.data.daos.impl

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory

import com.dkellycollins.triggerfinger.data.daos.IBitmapDao

import java.util.HashMap

class BitmapDao(private val _context: Context) : IBitmapDao {
    private val _store: HashMap<Int, Bitmap>

    init {
        _store = HashMap<Int, Bitmap>()
    }

    override fun loadBitmap(id: Int) {
        val bitmap = BitmapFactory.decodeResource(_context.resources, id)
        _store.put(id, bitmap)
    }

    override fun retrieve(id: Int): Bitmap {
        return _store[id]
    }

    override fun delete(id: Int) {
        val bitmap = _store[id]

        bitmap.recycle()

        _store.remove(id)
    }
}
